src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSPackageRestoreManager.cs

```diff
@@ -29,32 +31,49 @@ internal sealed class VSPackageRestoreManager : PackageRestoreManager
 
             SolutionManager = solutionManager;
             SolutionManager.NuGetProjectAdded += OnNuGetProjectAdded;
-            SolutionManager.SolutionOpened += OnSolutionOpenedOrClosed;
-            SolutionManager.SolutionClosed += OnSolutionOpenedOrClosed;
+            SolutionManager.SolutionOpened += OnSolutionOpened;
+            SolutionManager.SolutionClosed += OnSolutionClosed;
         }
 
-        private void OnSolutionOpenedOrClosed(object sender, EventArgs e)
+        private void OnSolutionOpened(object sender, EventArgs e)
+        {
+            // This is a solution event. Should be on the UI thread
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            var solutionDirectory = SolutionManager.SolutionDirectory;
+
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async delegate
+            {
+                await TaskScheduler.Default;
+                await RaisePackagesMissingEventForSolutionAsync(solutionDirectory, CancellationToken.None);
+            });
+        }
+
+        private void OnSolutionClosed(object sender, EventArgs e)
         {
             // This is a solution event. Should be on the UI thread
             ThreadHelper.ThrowIfNotOnUIThread();
 
             NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
-                {
-                    // We need to do the check even on Solution Closed because, let's say if the yellow Update bar
-                    // is showing and the user closes the solution; in that case, we want to hide the Update bar.
-                    var solutionDirectory = SolutionManager.SolutionDirectory;
-                    await RaisePackagesMissingEventForSolutionAsync(solutionDirectory, CancellationToken.None);
-                });
+            {
+                // We need to do the check even on Solution Closed because, let's say if the yellow Update bar
+                // is showing and the user closes the solution; in that case, we want to hide the Update bar.
+                var solutionDirectory = SolutionManager.SolutionDirectory;
+                await RaisePackagesMissingEventForSolutionAsync(solutionDirectory, CancellationToken.None);
```

Comment:
how about calling `PackagesMissingStatusChanged?.Invoke(this, new PackagesMissingStatusEventArgs(packagesMissing: false));` directly? Or create a new method `HidePackagesMissingStatusIfVisible`, so that the comment isn't needed and it's obvious it's efficiently hiding the status bar and not doing an expensive operation?

Reply:
That sounds like a good idea. 

I'll dig into it, but on a quick glance it makes sense.

Reply:
Implemented it. 
For the most part the PackagesMissingStatusChanged event handler seems to be null by the time the solution close event comes along, so it's mostly a no-op. Likely why we never saw any delays there. 

Not confident to completely change that though. 

With your suggestion we have saving some unneeded work.

Reply:
👍  The previous code called `RaisePackagesMissingEventForSolutionAsync` which made it sound like it was going to detect missing packages, which is pointless when closing the solution. Now it's clear that there isn't a performance bug without needing to read implementations of different methods.